baseplate/lib/secrets.py

```diff
@@ -464,6 +467,91 @@ def _get_file_data(self, filename: str) -> Tuple[Dict, float]:
         return result
 
 
+class VaultCSIEntry(NamedTuple):
+    mtime: float
+    data: Any
+    # A mutex to prevent possible duplicate work from event loops
+    updating: bool
+
+
+class VaultCSISecretsStore(SecretsStore):
+    """Access to secret tokens using a vault CSI mount with automatic refresh.
+
+    This store allows access to secrets stored in vault through the CSI interface.
+    It performs caching and will automatically reload secrets from disk. Generally
+    do not cache or store the values returned by this class's methods, rather get
+    them from this class each time you need them. The secrets are served from memory
+    when possible, so there's little performance impact in doing so, and you will be
+    sure to always have the current version in the face of key rotation etc.
+    """
+
+    path: Path
+    data_symlink: Path
+    cache: Dict[str, VaultCSIEntry]
+
+    def __init__(
+        self,
+        path: str,
+        parser: SecretParser,
+    ):  # pylint: disable=super-init-not-called
+        self.path = Path(path)
+        self.parser = parser
+        self.cache = {}
+        self.data_symlink = self.path.joinpath("..data")
+        if not self.path.is_dir():
+            raise ValueError(f"Expected {self.path} to be a directory.")
+        if not self.data_symlink.is_dir():
+            raise ValueError(
+                f"Expected {self.data_symlink} to be a directory. Verify {self.path} is the root of the Vault CSI mount."
+            )
+
+    def get_vault_url(self) -> str:
+        """Deprecated and will be removed in 3.0.0"""
+        raise NotImplementedError
+
+    def get_vault_token(self) -> str:
+        """Deprecated and will be removed in 3.0.0"""
+        raise NotImplementedError
+
+    def _get_mtime(self) -> float:
+        """Modification time is store-wide for CSI secrets."""
+        return os.path.getmtime(self.data_symlink)
+
+    def _raw_secret(self, name: str) -> Any:
+        try:
+            with open(self.data_symlink.joinpath(name), "r", encoding="UTF-8") as fp:
+                return self.parser(json.load(fp))
+        except FileNotFoundError:
+            # Try a second time in case the file was deleted while we tried to read it
+            try:
+                with open(self.data_symlink.joinpath(name), "r", encoding="UTF-8") as fp:
+                    return self.parser(json.load(fp))
+            except FileNotFoundError:
+                raise SecretNotFoundError(name)
```

Comment:
```suggestion
            except FileNotFoundError as exc:
                raise SecretNotFoundError(name) from exc
```

This'll let us have a nice exception and a useful stack trace